### PR TITLE
Small refactoring of #link_to_github logic

### PR DESCRIPTION
**Pull request by kerrizor**
I was writing up a script for a project on code metrics and spotted this method as one of the more complicated ones in the app. It wasn't /bad/ but I thought it could be a bit cleaner. This change drops its flog score from 38 to 16.4, but YMMV on if you think this is cleaner or not.

```
Before:
   158.6: flog total
     9.3: flog/method average

    38.0: RubygemsHelper#link_to_github    app/helpers/rubygems_helper.rb:18-25
    23.2: RubygemsHelper#links_to_owners   app/helpers/rubygems_helper.rb:91-95
    18.0: RubygemsHelper#subscribe_link    app/helpers/rubygems_helper.rb:44-58
    12.9: RubygemsHelper#simple_markup     app/helpers/rubygems_helper.rb:34-40
     9.5: RubygemsHelper#unsubscribe_link  app/helpers/rubygems_helper.rb:61-68

After:
   136.9: flog total
     8.1: flog/method average

    23.2: RubygemsHelper#links_to_owners   app/helpers/rubygems_helper.rb:91-95
    18.0: RubygemsHelper#subscribe_link    app/helpers/rubygems_helper.rb:44-58
    16.4: RubygemsHelper#link_to_github    app/helpers/rubygems_helper.rb:18-25
    12.9: RubygemsHelper#simple_markup     app/helpers/rubygems_helper.rb:34-40
     9.5: RubygemsHelper#unsubscribe_link  app/helpers/rubygems_helper.rb:61-68
     7.5: RubygemsHelper#latest_version_number app/helpers/rubygems_helper.rb:106-108
```

### CODE REVIEW

File: app/helpers/rubygems_helper.rb [hunk 1]
```diff
@@ -16,11 +16,13 @@ def link_to_page(id, url)
...
-      URI(@rubygem.linkset.code)
-    elsif !rubygem.linkset.home.nil? && URI(rubygem.linkset.home).host == "github.com"
-      URI(rubygem.linkset.home)
+    results = [rubygem.linkset.code, rubygem.linkset.home].map do |linkset|
+      uri = URI(linkset.to_s)
+
+      uri.host == "github.com" ? uri : nil
     end
+
+    results.reject(&:blank?).first
```
Comment (krainboltgreene) [inline app/helpers/rubygems_helper.rb]: Curious about the two steps here? Why assign a variable only to chain off of it anyways?

Comment (kerrizor) [reply app/helpers/rubygems_helper.rb]: I was grinding gears at 2am trying to get a clean `breaks and return` when we detect our first return in the loop, and gave up in the interest of "figure it out tomorrow". I **know** this is possible, I did it just a few weeks ago in another application, I just honestly can't recall for the life of me The One Weird Trick that makes it work! 😀 

..so the variable is really an artifact of struggling to make it work. I /could/ have chains the pruning of the returned array to the `map` block, but it just felt like I was Doing Too Much (too much "logistical density" about the block, if you follow my meaning). The variable assignment is a waste, but separates the 2 steps cleanly - refine to our possible values, then filter them for what we want.

(Appreciate the feedback, btw - this was really just a code exercise for myself last night, but its prompted some really interesting conversations today!)

Comment (krainboltgreene) [reply app/helpers/rubygems_helper.rb]: Ah, yes! That makes sense. I knew I'd get a good answer if I asked.

Were you looking for `Array.find()` by chance?

**Comment by jgaskins** [comment]
It's unfortunate that `Enumerable` doesn't have a "find and return first truthy value" because that's so useful. When I need to do that, I usually reach for `find` with a `break` inside it:

```ruby
    [rubygem.linkset.code, rubygem.linkset.home].find do |linkset|
      uri = URI(linkset.to_s)

      break uri if uri.host == "github.com"
    end
```

This also gets rid of intermediate array allocations from `map` and `reject` while also avoiding some `URI` allocations if it breaks early, but I don't know how much performance or memory that will actually save on a grand scale. Probably negligible.

But also, since breaking the loop also returns from the method, would it make sense to do `each` with an early `return` instead?

**Comment by smoak** [comment]
> It's unfortunate that Enumerable doesn't have a "find and return first truthy value" because that's so useful.

Perhaps I am misunderstanding the documentation, but I believe that is what `Enumerable#find` does?

> Passes each entry in enum to block. Returns the first for which block is not false.

"first for which block is not false" (aka true)

```
[rubygem.linkset.code, rubygem.linkset.home].map { |linkset| URI(linkset.to_s) }.find do |uri|
    uri.host == "github.com"
end
```

**Comment by jgaskins** [comment]
You're right, I wrote hastily and chose my words poorly.

What I meant was that it won't return the block result — it returns the element in the Enumerable instead. My interpretation of the PR description was that Kerri was looking for a way to do this without looping more than once:

> _Note that I /do/ think there's a cleaner way of returning from the loop on line 19 such that we can still honor the first match without having to `reject` and `first` on line 25_

`map.reject` loops 2x, `map.find` loops _at most_ 2x (which is an improvement in the average case), but `find { break }` (check the code I pasted) loops at most 1x. Technically, you could use `each { break }` since `break` causes `Enumerable` methods to return the value passed to `break`, but I use `find` because its name is closer to the intent.

All this is why I think there should be a method for this because that previous paragraph is a veritable _trash fire_.

**Comment by fschwahn** [comment]
You could use a lazy enumerator, which makes it a bit prettier in my opinion.

```ruby
[rubygem.linkset.code, rubygem.linkset.home].lazy
  .map { |linkset| URI(linkset.to_s) }
  .select { |uri| uri.host == "github.com" }
  .first
```

**Comment by kerrizor** [comment]
I can't believe I forgot about `#detect`.. which just goes to show you the richness of Ruby!

File: app/helpers/rubygems_helper.rb [hunk 2]
```diff
@@ -16,11 +16,11 @@ def link_to_page(id, url)
   end
 
   def link_to_github(rubygem)
-    if !rubygem.linkset.code.nil? && URI(rubygem.linkset.code).host == "github.com"
-      URI(@rubygem.linkset.code)
-    elsif !rubygem.linkset.home.nil? && URI(rubygem.linkset.home).host == "github.com"
-      URI(rubygem.linkset.home)
+    result = [rubygem.linkset.code, rubygem.linkset.home].detect do |linkset|
```
Comment (sonalkr132) [inline app/helpers/rubygems_helper.rb]: Looks much better :+1:
Can you please update it to use `Links`? You will need to pass `@latest_versions` as well from the view and use
`rubygem.links(version)` here to get links.

Comment (kerrizor) [reply app/helpers/rubygems_helper.rb]: Will do -- I'm holding on this only because it isn't terribly important to merge :)

**Comment by sonalkr132** [comment]
The function being changed has started using (metadata) links. Please let us know if you still plan to update this.

**Comment by simi** [comment]
What about https://github.com/rubygems/rubygems.org/pull/2980?

**Pull request closed without merging**